src/ParserWithComments.elm

```diff
@@ -0,0 +1,396 @@
+                                { comments = Rope.flatFromList (List.reverse commentsSoFar)
+                                , syntax = items
+                                }
+                            )
+                    )
+                ]
+    in
+    Core.loop ( [], [] ) manyWithoutReverseStep
+
+
+{-| Same as [`many`](#many), except that it doesn't reverse the list.
+This can be useful if you need to access the range of the last item.
+-}
+loop : a -> (a -> ParserWithComments (Core.Step a b)) -> ParserWithComments b
+loop initialState step =
+    let
+        stepWithComments :
+            ( List (Rope (Node String)), a )
+            ->
+                Core.Parser
+                    (Core.Step
+                        ( List (Rope (Node String)), a )
+                        { comments : Rope (Node String), syntax : b }
+                    )
+        stepWithComments ( commentsSoFar, state ) =
+            step state
+                |> Core.map
+                    (\pResult ->
+                        case pResult.syntax of
+                            Core.Loop pResultLoop ->
+                                Core.Loop
+                                    ( pResult.comments :: commentsSoFar
+                                    , pResultLoop
+                                    )
+
+                            Core.Done pResultDone ->
+                                Core.Done
+                                    { comments = Rope.flatFromList (List.reverse (pResult.comments :: commentsSoFar))
+                                    , syntax = pResultDone
+                                    }
+                    )
+    in
+    Core.loop ( [], initialState ) stepWithComments
+
+
+sepBy : String -> ParserWithComments a -> ParserWithComments (List a)
+sepBy sep p =
+    maybeMap identity
+        []
+        (sepBy1 sep p)
+
+
+sepBy1 : String -> ParserWithComments a -> ParserWithComments (List a)
+sepBy1 sep p =
+    map cons p
+        |> keep (many (Core.symbol sep |> Parser.Extra.continueWith p))
+
+
+sepBy1WithState : ParserWithComments () -> ParserWithComments a -> ParserWithComments (List a)
+sepBy1WithState sep p =
+    map cons p
+        |> keep
+            (many
+                (sep
+                    |> Core.andThen
+                        (\onlyComments ->
+                            p
+                                |> Core.map
+                                    (\pResult ->
+                                        { comments = onlyComments.comments |> Rope.prependTo pResult.comments
+                                        , syntax = pResult.syntax
+                                        }
+                                    )
+                        )
+                )
+            )
+
+
+between : Core.Parser () -> Core.Parser () -> ParserWithComments a -> ParserWithComments a
+between lp rp p =
+    (lp
+        |> Parser.Extra.continueWith p
+    )
+        |. rp
```

Comment:
(Just brainstorming)

These mix of `|.`/`|=` with `|>` require some odd parentheses a bit everywhere. I'm wondering whether we should instead use `|>` pipelines everywhere along with keep/ignore functions.

```elm
-- Maybe with a different name to not
keep = (|=)

foo
   |> keep someCoreParser
```

That way we can keep the constraint that `ignore` should work with `ParserWithComments ()` or `Parser ()`.

Reply:
`ParserWithComments ()` is not a thing. And we don't want to ignore comments anyway I don't think?
Anyway, could be reasonable. Maybe we just apply continueWith prefix-style to avoid the parens?

Reply:
Hmm, `Parser.WithComments ()` then? Anyway, what I meant is that prior to this PR, `Combine.ignore` took a `Combine.Parser State ()` as an argument, and that allowed us to make sure that the parser that was being ignored would not compute anything unnecessarily (like `Combine.string` back in 7.3.2). I think that's an attractive property to have if we can keep it.

(Note, I'm still not past the first comment, I'm just brainstorming here :sweat_smile: )